fdbserver/storageserver.actor.cpp

```diff
@@ -10652,6 +10660,13 @@ ACTOR Future<Void> storageServer(IKeyValueStore* persistentData,
 				recovered.send(Void());
 			return Void();
 		}
+		// Pass a reference of tenantPrefixIndex to the storage engine to support per-tenant data encryption,
+		// after the tenant map is recovered in restoreDurableState. In case of a storage server reboot,
+		// it is possible that the storage engine is still holding a pre-reboot tenantPrefixIndex, and use that
+		// for its own recovery, before we set the tenantPrefixIndex here.
+		if (self.encryptionKeyProvider.isValid()) {
+			self.encryptionKeyProvider->setTenantPrefixIndex(self.tenantPrefixIndex);
```

Comment:
how would this support usecase where `tenantPrefix` is not valid (`tenant_mode` == `optional` and/or `disabled`) ?

Reply:
The key provider will treat a tenant id as valid iff it is in the tenant prefix index. In `disabled` mode the tenant prefix index is empty, so the key provider will see all data being not belong to particular tenant. Likewise in `optional` mode only those tenant id found in the tenant prefix index will be treated as valid.